src/rust/engine/options/src/lib.rs

```diff
@@ -66,9 +57,56 @@ use crate::fromfile::FromfileExpander;
 use crate::parse::Parseable;
 pub use build_root::BuildRoot;
 pub use id::OptionId;
+use lazy_static::lazy_static;
+use parking_lot::Mutex;
 pub use scope::{GoalInfo, Scope};
+use serde::Deserialize;
+use std::any::Any;
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::fmt::Debug;
+use std::fs;
+use std::hash::Hash;
+use std::path::Path;
+use std::sync::Arc;
 pub use types::OptionType;
 
+lazy_static! {
+    static ref BIN_NAME: Mutex<String> = Mutex::new("pants".to_string());
+}
+
+// NB: This will be called at import time in several Python files to define static help strings
+// (e.g. "help=f'run `{bin_name()} fmt`"). So it must be set early. But on the other hand it can
+// only be set after we parse options and have access to the pants_bin_name option, which may
+// be too late: Some of the files that register options may already be using `bin_name()`.
+//
+// Fortunately, this only applies in pantsd; The native client does not have this problem because
+// it parses options without registration. So we use the __PANTS_BIN_NAME env var to propagate
+// this value from the native client to its spawned pantsd process (see pants_daemon_client.py).
```

Comment:
Ah, just confirming my understanding, the initialisation process looks something like:

1. the native client starts up, and parses options
2. parsing these options includes detecting the https://www.pantsbuild.org/stable/reference/global-options#pants_bin_name value (if set) and setting the static
3. `pants_daemon_client.py` then reads that option value via the static...
4. ... to propagate into pantsd via the env var

Reply:
Exactly right. This is in fact how it always worked, but the implementation obscured this. At a casual glance you could think that the daemon was reading the value and setting the env var. And this did happen, but too late. The reason it worked in practice was that the client read it first and then the spawned server received it via env var.

Reply:
Oh and I should mention that all this is tested (both with and without pantsd) in `tests/python/pants_test/integration/bin_name_integration_test.py`(https://github.com/pantsbuild/pants/blob/3e9d83061cbc1b043c9e409c9b2485df09e8de1a/tests/python/pants_test/integration/bin_name_integration_test.py)